### ISSUE TITLE
feat: Add execution time (`fit_time`/`score_time`) to the `EstimatorReport` and `CrossValidationReport` helpers

### ISSUE THREAD

**Issue by sylvaincom** [probabl-ai/skore#1241]
### Is your feature request related to a problem? Please describe.

Currently (skore v0.6.1), in the helpers of the `EstimatorReport` and the `CrossValidationReport`, we give access to metrics that contain scores, but no execution time.

As a data scientist, when assessing the performance of my model, I would like to have access to the score but also the execution time.

### Describe the solution you'd like

In the helpers, add the fit time and also the score time (prediction / inference).

The fit time should be straightforward. For the score time, we need to have at least one prediction computed to have access to it.

**Comment by MarieS-WiMLDS** [comment]
Can you precise where and how would like to see it in the report?

**Comment by sylvaincom** [comment]
I'd like to have them in the same dataframe obtained with `cv_report.metrics.report_metrics()` about metrics, but not sure what the best API for that would be. It kind of resembles a `cv_results_` from a scikit-learn grid search? Maybe `cv_report.cv_results()` because the time is not a metric per se?

<img width="450" alt="Image" src="https://github.com/user-attachments/assets/24952c63-e8a4-4d2b-a1de-3bcfb628abb9" />

**Comment by auguste-probabl** [comment]
Times are non-deterministic, so I hesitate to put them in the "metrics" category we have currently. For example, I don't think they should be put in the cache.

**Comment by sylvaincom** [comment]
If I remember correctly, @glemaitre believes the execution time is considered a metric and it would be fine to have them in `cv_report.metrics.report_metrics()`?

**Comment by GaelVaroquaux** [comment]
Hi everyone,
I considered using skore for the demo of skrub that I'm giving tomorrow, but I couldn't find an easy way of reporting time with very few lines of code (important to avoid loosing the audience). I'll stick to sklearn.model_selection.cross_validate for now, but I'm 👍  on this feature. It is important

**Comment by glemaitre** [comment]
Yep, we should find time introducing this feature back because it would unlock those analysis: https://github.com/probabl-ai/skore/issues/1340

@MarieSacksick Do you think that we could prioritize for the next sprint in 1.5 week?

**Comment by MarieSacksick** [comment]
We can! It means to postpone styling to make room for this, but I think styling is a nice to have.

**Comment by auguste-probabl** [comment]
Do I understand correctly that `score_time` should really be called `predict_time`? Otherwise, we'd have one `score_time` per metric (like `score_time_r2`, `score_time_rmse`...)

Here's my current idea:
- when the model is fitted, we measure the `fit_time` and put it in the cache. It will not be re-computed again.
- when the model is first evaluated, with any metric, we measure the `predict_time` (the time taken to run the model's `response_method`) and put it in the cache. It will not be re-computed again.

**Comment by auguste-probabl** [comment]
## Summary of a live review session of #1444 with @guillaume

### Design

- In the first iteration of #1444, querying the prediction time raises an exception if no metric has ever been computed. Instead, `report.metrics.predict_time()` should compute the predictions (and fill the cache, which speeds up later metric computations).

- Instead of two separate functions, `fit_time` and `predict_time`, we should have a single function, `timing`, with more parameters:

```python
def timing(
    kind: Literal["fit", "predict", "auto"] = "auto",
    data_source: DataSource = "test",
    X=None,
    y=None,
) -> dict:
    """Report fit and prediction times, in seconds.

    The times are reported in the form of a dict of the following shape:

    - If kind="fit", a single entry with key "fit"
    - If kind="predict", a single entry with key "predict_{data_source}" (where data_source
      is "train", "test" or "X_y_{data_source_hash}")
    - If kind="auto", two entries: a concatenation of the "fit" and "predict" cases.

    Fit information might not be available, e.g. if the estimator was fitted outside of the
    EstimatorReport. In this case, passing kind="fit" will result in an error, and passing
    kind="auto" is equivalent to passing kind="predict".
    
    Returns
    -------
    dict
        The times.

    Raises
    ------
    FitInformationUnavailableError
        If kind="fit" and the fit information is not available.

    Examples
    --------
    >>> report = ...

    >>> report.metrics.timing()
    {"fit": ..., "predict_test": ...}

    >>> report.metrics.timing(kind="fit")
    {"fit": ...}

    >>> report.metrics.timing(data_source="train")
    {"fit": ..., "predict_train": ...}

    >>> report.metrics.timing(data_source="X_y", X=..., y=...)
    {"fit": ..., "predict_X_y_njl2k34jk123s": ...}

    >>> report.metrics.timing(kind="predict", data_source="X_y", X=..., y=...)
    {"predict_X_y_njl2k34jk123s": ...}

    >>> report.metrics.timing(kind="fit", data_source="X_y", X=..., y=...)
    {"fit": ...}

    # Fit informtion is not available
    >>> report = ...

    >>> report.metrics.timing()
    {"predict_test": ...}

    >>> report.metrics.timing(kind="fit")
    FitInformationUnavailableError(...)
    """
```

By the way, why is it even possible to have an unfitted estimator?

### Implementation

In my first iteration of #1444 I put all the timing information in a separate place `report._timings_cache`. This implementation is more decoupled, which has benefits:
- Clearer responsibility of attributes
- Fewer tests to change (not gonna lie, this was a big reason)

But also issues:
- Cache can be invalid, e.g. in my first iteration `report.clear_cache` only cleared the "main" cache, because I didn't want the fit time to be deleted.

Right now, `report._cache` is the one big mess to put everything in (predictions, metrics...), so let's also put times in it. No real reason to refactor this now.
Make sure that `report.clear_cache` does not delete the fit time.

### Feedback needed

Names:
- `timing`?
- `kind`?

**Comment by MarieSacksick** [comment]
> In the first iteration of https://github.com/probabl-ai/skore/pull/1444, querying the prediction time raises an exception if no metric has ever been computed. Instead, report.metrics.predict_time() should compute the predictions (and fill the cache, which speeds up later metric computations).

About this, I disagree. If prediction takes a while (like, more than a couple of seconds), I'm going to be surprised as a user: I asked to know the timing, not to compute. It's not a strong disagreement though, if I have a message saying something like `Prediction has ever been computed for this dataset, doing it to be able to render the prediction time`, then it's explicit and I wouldn't be surprised.  

About the function `timing`, I don't find it clear. I liked that fit and predict had two different functions, because for fit it's once and for all, while for predict it depends on the dataset. What happens for `estimator.metrics.timing(kind = "fit", X = "test")` for instance? Is it going to fit again my estimator on X_test? This would be completely out of the scope of this function.

**Comment by glemaitre** [comment]
> About this, I disagree. If prediction takes a while (like, more than a couple of seconds), I'm going to be surprised as a user: I asked to know the timing, not to compute. It's not a strong disagreement though, if I have a message saying something like Prediction has ever been computed for this dataset, doing it to be able to render the prediction time, then it's explicit and I wouldn't be surprised (just maybe a bit annoyed that something takes a while).

This would call the `cache_prediction` function and thus we could add indeed a message that we need to collect the info.

**Comment by glemaitre** [comment]
> What happens for estimator.metrics.timing(kind = "fit", X = "test") for instance?

`data_source` is ignored for for `kind="fit"`

The fact that I need to make three call for:

```python
report.metrics.fit_time()
report.metrics.predict_time(data_source="train")
report.metrics.predict_time(data_source="test")
```

is a bit annoying (knowing that we would need `data_source="all"` for actually having a single call).

The current `cross_validate` in scikit-learn would return you fit/predict and I would expect people to want the same info. Usually, you are looking at the trade-off fit/predict and not only one of them.

> Is it going to fit again my estimator on X_test?

We will never refit. If you pass an unfitted estimator with and request `"fit"`, we would raise an error. It is the reason to have the `"auto"` to be permissive in case you want the timing but don't want to be bother.

**Comment by glemaitre** [comment]
I'm fine with the name `kind`.

**Comment by MarieSacksick** [comment]
> The fact that I need to make three call [...]

I get the pain. But I'm not convinced that returning a dict actually fixes it, because I have then to access the key I'm interested in.  
I'm not sure we want to look at the trade-off fit/predict, because in general they are positively correlated. I'd argue that depending on the use-case, I'm more interested in one or the other.  

What about adding an option to `data_source`, "both", so we would only have 2 lines:  

```
report.metrics.fit_time()
report.metrics.predict_time(data_source="both")
```

**Comment by glemaitre** [comment]
> I get the pain. But I'm not convinced that returning a dict actually fixes it, because I have then to access the key I'm interested in.

Right now, in terms of API, we let the individual "metrics" function to return Python native type (dictionary, float, etc.)

If you are interested in the dataframe feels, then it is the job of `report_metrics()`:

```python
report.metrics.report_metrics(scoring="timing")
```

that internally will call the `timing` function and foward the `data_source` and concatenate the info.

> I'm not sure we want to look at the trade-off fit/predict, because in general they are positively correlated. I'd argue that depending on the use-case, I'm more interested in one or the other.

Until I have a parameter to control it like `kind`, I don't see the issue.

> What about adding an option to data_source, "both", so we would only have 2 lines:

Yes, but I want to have it consistent for all metrics at once.
